### DISCUSSION TITLE
Release Candidate v4.0.0rc

### DISCUSSION THREAD

**Discussion by dewenni** [dewenni/ESP_Buderus_KM271#80] [Announcements]
Hi together!

It's been a bit quiet recently, but something has been happening in the background.

## what's new

This major Update brings a completely new web frontend 🎉
Instead of previous that was based on ESPUI-Library, the new one is build on basic html, css and javascript.
Also the communication between server and client has changed from websocket to SSE (Server-Sent Events) and Fetch-API. 

The new webUI Desktop style:
<img width="800" alt="weubui_dash" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/55fdd856-45bd-4f1f-9875-da9aaad8db78">
(Desktop Version)

and in fact that the new WebUI has responsive design, it offers also a mobile layout:

<img width="800" alt="image" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/a551c0cf-1fd2-48ce-afd6-d8ff7a716219">
(Mobile version)

### Messaging option - Pushover

There is now the possibility to get messages via Pushover. 
Depending on the parameter "Filter", you can define what kind of messages you want to receive.
In the settings you can find all necessary parameters to setup the client.

Each application, service, or utility that sends notifications through Pushover's API needs to have its own API token which uniquely identifies all of the API requests that it makes.
API tokens are free and can be registered through [Pushover website](https://pushover.net/apps/build).

<img width="800" alt="pushover" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/a554d0e2-5682-4c04-8c32-a72b7253e27b">

### WebUI-Logger

There is also a new log function with which you can record various messages depending on the filter and display them via the WebUI. This can be useful for your own debugging and also for the further development of the software.

<img width="800" alt="logger" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/e98ea323-a00a-4122-882d-0f302f48e7c4">

### Telnet

In addition to the WebUI and MQTT, there is also a new Telnet interface to communicate with the ESP.
The interface offers several commands to read out information and send commands.
An overview of the commands can be called up using the "help" command.
To connect, a simple Telnet connection can be started via the corresponding IP address of the ESP.

Example: 
```
> telnet 192.168.178.135
```
<img width="800" alt="telnet" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/6aa60b90-4a1c-49fa-9640-ffc2bf23649f">

## changelog

- complete new web UI.
- removed ESPAsyncWebServer and AsyncTCP. Instead I use [PsychicHttp](https://github.com/hoeken/PsychicHttp) that seems to be more stable and even faster.
- OTA and Config import/export was redesigned
- login authentication was redesigned
- rework messaging function. mqtt publish function was replaced with a generic messaging function for further options.
- add messaging option for "Pushover" #31
- separate mqtt language option independent of webUI language
- add WebUI-Logger #70
- bugfix for updating OneWire sensor values if not enabled
- Prevent zero values from being sent at startup for "info = {"burner":0,"pump":100,"ww_temp":57,"boiler_temp":38}"
- improve behavior for incoming mqtt messages without payload (this could lead to problems in the past)
- change DoubleResetDetection time from 10s to 2s
- local copy of ESPAsyncWebServer with merged PR887 and PR1211
- add SIM_MODE for testing purposes (can be activated in config.h)

> [!IMPORTANT]   
> With this version, the dependency of the WebUI language and the MQTT messages has been removed. There is now a separate language setting in the MQTT settings that is used for the MQTT topics!

## Download
UPDATE 30.03.2024 - 12:05
[buderus_km271_v4.0.9002.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/14811615/buderus_km271_v4.0.9002.zip)

**Comment by dewenni** [comment]
For all those interested, I have a preview version with new functions for you here. I would like to use the discussion here to get feedback on the new functions before I make an official release.
Feel free to test it out and give me feedback if possible (in German or English).

---

Für alle Interessierten habe ich hier für euch eine Vorabversion mit neuen Funktionen. Ich möchte den Weg über die Diskussion hier nutzen um Feedback zu den neuen Funktionen zu bekommen, bevor ich ein offizielles Release mache.
Testet es gerne aus und gebt mir wenn möglich Feedback dazu (deutsch oder englisch).

**Comment by dewenni** [comment]
if someone already testing, I have updated the bin files, because I found a problem regarding browser refresh.

**Reply by JieffeEl** [reply]
I quickly checked the new release on a stand-alone ESP32 but it seems there is a problem with the language selection. When I switch to English in the setup ("Englisch" is a typo in your program BTW), the program restarts to the setup portal 192.168.4.1.

**Reply by dewenni** [reply]
hey @JieffeEl 
thanks for testing. What you observe is new for me. I have checked it again on two different systems and it works fine for me.
This language switching is also just a client function. Almost every static text in the WebUI is in a javascript file and when you change the language, the javascript function replaces all texts according to the selected language.

Have you used a config.json from an existing device or have you tested with an empty configuration and only edited the WiFi parameters?

btw. "Englisch" is not really a typo. It is the german word. If you set it to German, both parameters are in german.
If you switch to English, both parameters are in english.

![change_language](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/01e814f5-544d-410d-bb7d-de20e8c25255)

**Reply by JieffeEl** [reply]
Okay, the problem is somewhere else, the watchdog restarts the ESP ... 

Info on the board I'm using for testing : 

Detecting chip type... ESP32
Chip is ESP32-D0WDQ6 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz

I first erased the ESP32, then flashed the .BIN. I connected through the portal while keeping an eye on the progress via a serial port terminal.

The watchdog is triggered within a few seconds. See the log attached.

I'll try to find another ESP32 to test (not doing the testing on the actual interface card until the cold season is over, just to be on the safe side :) )

[buderus.log](https://github.com/dewenni/ESP_Buderus_KM271/files/14802980/buderus.log)

**Reply by dewenni** [reply]
It is caused by async_tcp.
The question why?!

E (29530) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (29530) task_wdt:  - async_tcp (CPU 0/1)

**Reply by JieffeEl** [reply]
Tried with 2 different ESP32, same result. Tried using Firefox and Chrome (I'm on Linux Debian, but I don't think it matters) to access the portal, same result too.

Edit : Just to be sure, I tested from a Windows 10 VM with Firefox, still rebooting

**Reply by JieffeEl** [reply]
Could it come from the way I flashed the .BIN ? My esptool command line : 
`
python -m esptool --chip esp32 --port /dev/ttyUSB0 write_flash -z 0x0 buderus_km271_esp32_flash_v4.0.0rc.bin 
`

**Reply by dewenni** [reply]
I usually use this: `esptool.py -p <UPLOAD-PORT> write_flash 0x00 buderus_km271_esp32_flash_vx.x.x.bin `
But I don't think that it is related to the flash process.
I'm going to do some research and maybe switch to different AsyncTCP and EspAsyncWebServer Libraries.

**Reply by dewenni** [reply]
@JieffeEl 
I have done some redesign and have removed ESPAsyncWebServer ans AsyncTCP. Instead I use [PsychicHttp](https://github.com/hoeken/PsychicHttp) that seems to be more stable and even faster.

I am happy if you try this new version 4.0.9002  that I have updates in the initial post.

Lets see if this is a good alternative, otherwise I will try this one: [H4AsyncWebServer](https://github.com/HamzaHajeir/H4AsyncWebServer)

**Reply by JieffeEl** [reply]
Working now, more or less ! 

Connects to the wifi and MQTT without problem on first reboot.

The interface reverts to German on first connection and when I do a refresh in the browser. It then returns to English after around 10 seconds. It might come from the fact that my test board is not connected to the boiler and thus a delay is introduced while the program waits for (non-existent) data.

However, when going back to "Settings" after connection to the network, the "WIFI" tab is empty. On next boot, it's not reconnecting :(

While I don't go to "Settings" after the initial setup, the board reconnects after each reboot. Seems the wifi credentials are erased when accessing the "Settings" tab. Doesn't happen to the other settings, like MQTT ...

///////////////////////
Log : 

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13964
load:0x40080400,len:3600
entry 0x400805f0
LittleFS Flag read = 0xD0D04321
No doubleResetDetected
Saving config file...
Saving config file OK
LittleFS successfully started
[   232][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to: 
Telnet Server: running!
Stop doubleResetDetecting
Saving config file...
Saving config file OK
[ 10275][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to:  - attempt: 1/5
[ 20276][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to:  - attempt: 2/5
[ 30278][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to:  - attempt: 3/5
[ 40280][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to:  - attempt: 4/5
[ 50282][E][WiFiSTA.cpp:232] begin(): SSID too long or missing!
WiFi Mode STA - Trying connect to:  - attempt: 5/5

! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! !

Wifi connection not possible, esp rebooting...

! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! ! !`
//////////////

Anyway, I like the new interface. That's a splendid job you've made so far.

**Reply by dewenni** [reply]
thanks again!

I also see that behavior with the WiFi settings, that sometimes not updates in the settings.
But the handling is the same as the other ones - strange....

This web server stuff is really "bitchy" on an ESP.
The ESPAsyncWebServer and the AsyncTCP from me-no-dev are good libraries, but there are also some known issues that are not solved. If you stress it too much, the ESP tends to crash.
That is why I want to try an alternative to that.

I think I will also give this one [H4AsyncWebServer](https://github.com/HamzaHajeir/H4AsyncWebServer) a try and see what is better.

and yes, if you are not connected to the Logamatic, it could be that you have a small deadlock a the beginning until the serial connection gets time outed.

**Comment by dewenni** [comment]
I have now published release 4.0.0

For further discussions please visit #82 
For specific bugs please create a new issues
